asllib/doc/ASLSemanticsReference.tex

```diff
@@ -19,160 +283,1634 @@
+
+We define two types of semantics --- a \emph{sequential semantics} and a \emph{concurrent semantics}.
+
+\paragraph{Concurrent Semantics.}
+The concurrent semantics includes \herd\ execution graphs.
+Intuitively, these graphs define read and write effects to variables and constraints over those effects.
+Together with the constraints that define the ARM memory model~\cite{AlglaveDGHM21}, these graphs axiomatically define
+the valid sequences of read and write operations induced by a given specification.
+
+\paragraph{Sequential Semantics.}
+The sequential semantics corresponds to executing an ASL specification in the context of a single (conceptual) process
+of execution --- notice that ASL does not contain any concurrency constructs. Access to memory is hidden away
+in primitive subprograms that interface with a map-like memory implementation that does not reorder memory accesses
+and serves one access at a time.
+%
+Technically, the sequential semantics is defined by dropping the \herd\ execution graph components
+from the configurations.
+
+\section{Mathematical Notations and Conventions}
+
+\input{ASLNotations.tex}
+
+\paragraph{OCaml-style Notations.}
+We use the notation $L(v_{1..k})$, where $L$ is a label and $v_{1..k}$ is a (possibly-singleton) tuple of abstract values,
+to denote the tuple $(L,v_{1..k})$.
+We also write $L(T_{1..k})$, where $T_{1..k}$ denote mathematical type of values, to stand for
+$\{L\} \times T_1 \times \ldots \times T_k$.
+%
+We later use these notations to define \emph{semantic configurations} and \emph{native values}, \underline{not to be confused with labelled AST nodes}.
+%
+The notation $\langle \cdot \rangle$ stands for empty-or-singleton sets, where $\None\triangleq\langle\rangle$ denotes
+an empty set and $\langle v \rangle$ denotes a set containing the single element $v$.
+%
+The notation $\langle T \rangle$, where $T$ denotes a mathematical type of values, stands for
+$\{ \langle\rangle \} \cup \{\langle v \rangle \;|\; v \in T\}$.
+
+\paragraph{Misclleaneous.}
+We define the following shorthand notations for the true and false expressions:
+\newcommand\etrue[0]{\texttt{ETrue}}
+\newcommand\efalse[0]{\texttt{EFalse}}
+\[
+\begin{array}{rcl}
+  \etrue &\triangleq& \texttt{E\_Literal}(\texttt{L\_Bool}(\texttt{True}))\\
+  \efalse &\triangleq& \texttt{E\_Literal}(\texttt{L\_Bool}(\texttt{False}))\\
+\end{array}
+\]
+
+\section{Configurations}
+
+Configurations express intermediate states related by \emph{semantic relations}.
+%
+More precisely, semantic relations relate two distinct sets of configurations ---
+\emph{input configurations} and \emph{output configurations}.
+Input configurations consist of an environment and an AST node.
+Output configurations consist of an output environment, values,
+and \herd\ execution graphs.
+%
+Configurations wrap together elements such as environments and AST nodes
+and associate them with a \emph{mode}. Input configurations use modes
```

Comment:
could "mode" be called something like "rule application domain" or "rule use case"? (throughout)

Reply:
Those are too long. How about label? The configuration/state label?

Reply:
I don't think so - please use long names and if you cannot be bothered typing long names, please use a macro that will expand to the long name. Most of our audience needs our help for readability.

Reply:
The thing is modes also label output configurations so they're not necessarily related to rules.
Perhaps "configuration types"?

Reply:
configuration domain?